composer/trainer/trainer.py

```diff
@@ -680,25 +685,44 @@ def _train_loop(self) -> None:
+                    rerun_train_batch = True
                     total_loss = None
-                    microbatches = self._train_data_spec.split_batch(state.batch, state.grad_accum)
-                    if self.deepspeed_enabled:
-                        total_loss = self._train_batch(microbatches)
-                    elif self._use_closures():
-                        for optimizer in state.optimizers:
-                            if use_grad_scaling:
-                                total_loss = state.scaler.step(
-                                    optimizer, closure=lambda **kwargs: self._train_batch(microbatches, **kwargs))
+                    while rerun_train_batch:
+                        try:
+                            rerun_train_batch = False
+                            microbatches = self._train_data_spec.split_batch(state.batch, state.grad_accum)
+                            if self.deepspeed_enabled:
+                                total_loss = self._train_batch(microbatches)
+                            elif self._use_closures():
+                                for optimizer in state.optimizers:
+                                    if use_grad_scaling:
+                                        total_loss = state.scaler.step(
+                                            optimizer,
+                                            closure=lambda **kwargs: self._train_batch(microbatches, **kwargs))
+                                    else:
+                                        total_loss = optimizer.step(
+                                            closure=lambda **kwargs: self._train_batch(microbatches, **kwargs).item())
                             else:
-                                total_loss = optimizer.step(
-                                    closure=lambda **kwargs: self._train_batch(microbatches, **kwargs).item())
-                    else:
-                        total_loss = self._train_batch(microbatches)
-                        for optimizer in state.optimizers:
-                            if use_grad_scaling:
-                                state.scaler.step(optimizer)
+                                total_loss = self._train_batch(microbatches)
+                                for optimizer in state.optimizers:
+                                    if use_grad_scaling:
+                                        state.scaler.step(optimizer)
+                                    else:
+                                        optimizer.step()
+                        except RuntimeError as e:
+                            if self.adaptive_grad_accum and "CUDA out of memory" in str(e):
+                                rerun_train_batch = True
+                                # Raise runtime error if we can't train even one sample at a time
+                                if state.grad_accum >= num_samples_in_batch:
```

Comment:
`num_samples_in_batch` is the total across GPUs. I think you want `state.batch_num_samples` instead.